--- a/avahi-core/server.c
+++ b/avahi-core/server.c
@@ -676,11 +676,37 @@ static void handle_response_packet(AvahiServer *s, AvahiDnsPacket *p, AvahiInterface *i, const AvahiAddress *a, int from_local_iface) {
             break;
         }

-        if (!avahi_key_is_pattern(record->key)) {
-            /* Filter services that will be cached. Allow all local services */
-            if (!from_local_iface && s->config.enable_reflector && s->config.reflect_filters != NULL) {
-               AvahiStringList *l;
-               int match = 0;
+
+        int isfilterenabled = 0;
+
+        if (!avahi_key_is_pattern(record->key) && !from_local_iface && s->config.enable_reflector && s->config.reflect_filters != NULL) {
+            AvahiStringList *l;
+
+            isfilterenabled = 1;
+
+            for (l = s->config.reflect_filters; l; l = l->next) {
+                if (record->key->type == AVAHI_DNS_TYPE_PTR) {
+                    char* match = strstr(record->data.ptr.name, (char*) l->text);
+                    if (match != NULL) {
+                        avahi_log_debug("Reject Ptr SRC [%s] - [%s] iface [%d]", match, record->data.ptr.name, from_local_iface);
+                        goto unref;
+                    }
+                }
+                else if (record->key->type == AVAHI_DNS_TYPE_SRV || record->key->type == AVAHI_DNS_TYPE_TXT) {
+                    char* match = strstr(record->key->name, (char*) l->text);
+                    if (match != NULL) {
+                        avahi_log_debug("Reject Key match [%s] - [%s] iface [%d]", match, record->key->name, from_local_iface);
+                        goto unref;
+                    }
+                }
+            }
+        }
+
+        if (!avahi_key_is_pattern(record->key)) {
+            /* Filter services that will be cached. Allow all local services */
+            if (!isfilterenabled && !from_local_iface && s->config.enable_reflector && s->config.reflect_filters != NULL) {
+                AvahiStringList *l;
+                int match = 0;

                 if (record->key->type == AVAHI_DNS_TYPE_PTR) {
                     /* Need to match DNS pointer target with filter */
